### PR TITLE
fix: Incorrect Dart and Flutter SDKs compatibility range

### PR DESCRIPTION
**Pull request by Nidal-Bakir**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

### Issue Description

Closes: #872 

### Approach

Supporting older versions of the framework and language is not ideal because we miss out on new additional features. Additionally, if the Flutter team removes deprecated APIs (as they did in 3.9.x, see issue #861), it can cause problems for us because we still support an older version that had the deprecated API, while also needing to support a newer version that no longer has the API.

### CODE REVIEW

**Comment by Nidal-Bakir** [comment]
Do not merge until the Changelog entry `5.0.0` is modified to indicate the bump of the min SDK version.
I will edit the Changelog entry `5.0.0` when #860 merged so I can see the entry in this PR

**Comment by mtrezza** [comment]
What's the purpose of this PR? It doesn't seem to be a breaking change, it's only an upgrade from `>=2.12.0 <3.0.0` to
`>=2.17.0 <3.0.0`.

**Comment by Nidal-Bakir** [comment]
See the discussion

> If we remove a property and therefore remove support for currently supported Flutter SDK versions then this is a breaking change. Let's discuss this [in the issue](https://github.com/parse-community/Parse-SDK-Flutter/issues/861#issuecomment-1483899429)

**Comment by mbfakourii** [comment]
@Nidal-Bakir 
I don't think it is a breaking change either
Are you sure about this?

**Comment by mtrezza** [comment]
@Nidal-Bakir You are correct with your [comment](https://github.com/parse-community/Parse-SDK-Flutter/pull/867#issuecomment-1518674540) that changing the lower bound would be a breaking change. But what is the purpose of this PR, why this lower bound change?

**Comment by Nidal-Bakir** [comment]
> But what is the purpose of this PR

I'm using a pull request to make this change because I don't see a good place to put it. We cannot make such a change with other pull requests. For example, we cannot bump the minimum required Dart SDK in this pull request (https://github.com/parse-community/Parse-SDK-Flutter/pull/860) because it would not be noticeable in the commit log. Therefore, I opened a new pull request specifically for this bump so that we can see it in the commit log

**Comment by Nidal-Bakir** [comment]
> I don't think it is a breaking change either
Are you sure about this?


https://github.com/parse-community/Parse-SDK-Flutter/issues/861#issuecomment-1483899429
Please read the comment I am referring to, as well as the comments that follow it.

**Comment by mtrezza** [comment]
I remember now, thanks. I think we need to setup a strategy for when remove support for a dart or flutter version.

Usually we follow the long-term support schedule of the framework publisher, but I couldn't find any infos about that for Dart / Flutter.

**Comment by Nidal-Bakir** [comment]
> I remember now, thanks.

You are most welcome 

> I think we need to find a strategy for when remove support for a dart or flutter version.

Basically, we do not need any special strategy for bumping the Dart SDK constraint. We simply bump the constraint like any other package. For example, take a look at this PR on the flutter-fire repository: [feat: bump dart SDK constraint to 2.18](https://github.com/firebase/flutterfire/pull/10618). They did not even count it as a breaking change because, in the Flutter ecosystem, we do not have the concept of LTS or worry about any app that uses a Dart version older than the current stable one. This is because you should have migrated to a newer version a long time ago.


> Usually we follow the long-term support schedule of the framework publisher, but I couldn't find any infos about that for Dart / Flutter.

Flutter does not have the concept of LTS you should always migrate to the latest stable release

**Comment by mtrezza** [comment]
I've opened https://github.com/parse-community/Parse-SDK-Flutter/issues/872 to discuss this in more detail.

**Comment by mtrezza** [comment]
Before we merge this we should resolve https://github.com/parse-community/Parse-SDK-Flutter/issues/872#issuecomment-1541916230, because it seems like the upper bound may need to be changed as well. So let's wait for the outcome of the discussion there.

File: packages/dart/pubspec.yaml
Comment (Nidal-Bakir) [reply packages/dart/pubspec.yaml:7]: After dart 3 Now this should be like this.

sdk: ">=2.18.0 <4.0.0"

Comment (Nidal-Bakir) [reply packages/dart/pubspec.yaml:7]: this "<2.20.0" will prevent any dart 3 users from using the SDK

Comment (mtrezza) [reply packages/dart/pubspec.yaml:7]: With https://github.com/parse-community/Parse-SDK-Flutter/pull/874 the CI tests also with Dart 3, so I guess we can say `>=2.18.0 <3.1.0`. What's really strange is that [publish dry-run](https://github.com/parse-community/Parse-SDK-Flutter/actions/runs/4953846896/jobs/8861768216?pr=874) says:

> * The declared SDK constraint is '>=2.12.0 <3.0.0', this is interpreted as '>=2.12.0 <4.0.0'.
>   
>   Consider updating the SDK constraint to:
>   
>   environment:
>     sdk: '>=2.12.0 <4.0.0'

Comment (Nidal-Bakir) [reply packages/dart/pubspec.yaml:7]: I'm really confused, just like you, especially when I read "`interpreted as '>=2.12.0 <4.0.0'.`" It made me second-guess all my decisions!

> With https://github.com/parse-community/Parse-SDK-Flutter/pull/874 the CI tests also with Dart 3, so I guess we can say >=2.18.0 <3.1.0

`>=2.18.0` probably will not work https://github.com/parse-community/Parse-SDK-Flutter/pull/874#issuecomment-1544841675

Comment (mtrezza) [reply packages/dart/pubspec.yaml:7]: You wrote:

> Flutter 3.10.0 and Dart 3 require full sound null safety but the examples are not null safe

What are these "examples"?

Comment (Nidal-Bakir) [reply packages/dart/pubspec.yaml:7]: The example folder in the flutter package

Comment (mtrezza) [reply packages/dart/pubspec.yaml:7]: The examples and the test file using the `exclusive` argument should not force the SDK to reduce its compatibility. They are not really part of the SDK after all.

Can we remove the parameter 'exclusive' from the test code?

Comment (Nidal-Bakir) [reply packages/dart/pubspec.yaml:7]: > Can we remove the parameter 'exclusive' from the test code and then use the <=2.18.0 range in this PR?

Do you mean we can use Flutter >=3.3.0? 

if so then yes we can I just removed the `exclusive` parameter and currently, we officially support Flutter >=3.3.0 <=3.7.12

See #876

Comment (mbfakourii) [reply packages/dart/pubspec.yaml:7]: I think it's better to update it, it might have a bug, it hasn't been checked for a long time

**Comment by mtrezza** [comment]
I've removed the Dart <2.18 jobs from the CI, since the Parse Dart SDK is only compatible from >= 2.18

File: packages/dart/pubspec.yaml
Comment (mtrezza) [reply packages/dart/pubspec.yaml]: I think the issue with moving the examples out of the package is that they will be hard to discover for developers. The won't be published and will only be available in the GitHub repository. If that's currently needed, we may do that and re-add them later, but if we can keep them in the Flutter package easily we should prefer that I think.

Comment (Nidal-Bakir) [reply packages/dart/pubspec.yaml]: You're right about this. However, as a temporary solution, we can move them out of the Flutter package and deal with them later.

Comment (mtrezza) [reply packages/dart/pubspec.yaml]: I've moved it to the root and added a link to the README of the Flutter package.

**Comment by mtrezza** [comment]
I've added a flutter range, and the CI failed with:

> Package validation found the following potential issue:
> * The Flutter constraint should not have an upper bound.
>   In your pubspec.yaml the constraint is currently `>=3.3.0 <3.11.0`.
>   
>   You can replace that with just the lower bound: `>=3.3.0`.

That's a truly interesting concept.

**Comment by mtrezza** [comment]
Only Flutter 3.10 fails with:

>   info • 'window' is deprecated and shouldn't be used. Look up the current FlutterView from the context via View.of(context) or consult the PlatformDispatcher directly instead. Deprecated to prepare for the upcoming multi-window support. This feature was deprecated after v3.7.0-32.0.pre • packages/flutter/lib/parse_server_sdk_flutter.dart:81:32 • deprecated_member_use
>   error • The named parameter 'vsync' isn't defined • packages/flutter/lib/src/utils/parse_live_list.dart:276:9 • undefined_named_parameter

**Comment by Nidal-Bakir** [comment]
The Dart SDK is ready:
We need to merge the PRs to release Dart 5.0.0
- This PR
- https://github.com/parse-community/Parse-SDK-Flutter/pull/860

File: packages/flutter/lib/parse_server_sdk_flutter.dart
Comment (mtrezza) [inline packages/flutter/lib/parse_server_sdk_flutter.dart:81]: What is this change?

**Review by mtrezza — CHANGES_REQUESTED** [review]
Could you add a changelog entry?

File: packages/flutter/lib/parse_server_sdk_flutter.dart
Comment (Nidal-Bakir) [reply packages/flutter/lib/parse_server_sdk_flutter.dart:81]: > info • 'window' is deprecated and shouldn't be used. Look up the current FlutterView from the context via View.of(context) or consult the PlatformDispatcher directly instead. Deprecated to prepare for the upcoming multi-window support. This feature was deprecated after v3.7.0-32.0.pre • packages/flutter/lib/parse_server_sdk_flutter.dart:81:32 • deprecated_member_use
error

The `ui.window.locale` is deprecated. Should I open a new PR to describe this change?

**Comment by Nidal-Bakir** [comment]
> Could you add a changelog entry?

Sorry, but I am a bit confused about what should I put it. Should we merge this #860 first and then update the entry?

Since we need 2 PRs to release Dart 5.0.0. 

https://github.com/parse-community/Parse-SDK-Flutter/pull/860#pullrequestreview-1415937698
> * Since this won't be released, there is no version 5.0.0, so it also can't be in the changelog yet.
> * Reformatted the BREAKING CHANGES section to align with org-wide style.

File: packages/flutter/lib/parse_server_sdk_flutter.dart
Comment (mbfakourii) [reply packages/flutter/lib/parse_server_sdk_flutter.dart]: @Nidal-Bakir 
Its replacement has been changed as follows
```ui.window.locale``` to ```ui.PlatformDispatcher.instance.locale```

**Comment by mtrezza** [comment]
We don't have auto-release yet, which would generate a changelog entry for each PR according to the commit message (=PR title). So what we do instead until when is with each PR add a changelog entry for the PR, just like it would be auto-generated.

So currently, the changelog entry would be the PR title, and if necessary - in addition - an entry in the BREAKING CHANGES section.

File: packages/flutter/lib/parse_server_sdk_flutter.dart
Comment (mtrezza) [reply packages/flutter/lib/parse_server_sdk_flutter.dart]: Got it, thanks

**Comment by mtrezza** [comment]
I've added the changelog entries for dart and flutter. This is kind of a rarety, as we have 1 PR that creates changelog entry in both packages; I think that usually wouldn't work this way with auto release, but we can do it here.

**Comment by Nidal-Bakir** [comment]
Thanks for clarifying.

**Comment by mtrezza** [comment]
I think we can merge this?

**Comment by Nidal-Bakir** [comment]
Yes, we are ready.

**Pull request merged**
